api.go

```diff
@@ -606,6 +611,8 @@ func NewRaft(conf *Config, fsm FSM, logs LogStore, stable StableStore, snaps Sna
 	// to be called concurrently with a blocking RPC.
 	trans.SetHeartbeatHandler(r.processHeartbeat)
 
+	r.recoverFromCommitedLogs()
```

Comment:
There is a loop above that starts with:

```go
         // Scan through the log for any configuration change entries.
	snapshotIndex, _ := r.getLastSnapshot()
	for index := snapshotIndex + 1; index <= lastLog.Index; index++ {
```
(which GH doesn't let me comment on directly 🙄 ) that I wonder if we should incorporate somehow 🤔 .

It was already iterating the whole log to see if any logs are config changes so it would be nice to incorporate into the same loop somehow if we are restoring rather than read and deserialise every log entry twice. In fact `processLogs` will already do this config stuff to if we do call `recoverFromCommittedLogs`!

Also: 

```suggestion
	r.recoverFromCommittedLogs()
```
nit: but if we keep this name we should correct the typo in committed.

----

Finally, in writing the above I've realised that he config change loop in the existing code is possibly wrong and maybe even violate's Raft's correctness 😱 . I need to think about it some more but the whole reason we are not able to process all the logs through the FSM yet is because we don't know for sure if there were committed, yet in the loop above we are just blindly calling `processConfigurationLogEntry` on anything we find which assumes the log is committed and sets `lastCommittedConfiguration` to whatever you pass in!

I'm going to create a separate issue for this. Let's leave the behaviour unchanged for now here when not using fast recovery at least.

Reply:
On the last part. I spent a while thinking about the correctness of this and I think it's OK as it is 😅 .

We implement the "simplified" membership change protocol form the thesis (chapter 4) where only single node addition or removal are allowed at once and where we only allow a single uncommitted membership change to be in flight. Under these invariants, we don't need "joint consensus" because every quorum in both old and new configurations must still overlap. That means that in normal case, servers adopt new configurations immediately even before they are known to be committed and this is safe by that given logic:

>  If servers adopted Cnew only when they
learned that Cnew was committed, Raft leaders would have a difficult time knowing when a majority of the old cluster had adopted it. They would need to track which servers know of the entry’s
commitment, and the servers would need to persist their commit index to disk; neither of these
mechanisms is required in Raft. Instead, each server adopts Cnew as soon as that entry exists in its
log, and the leader knows it’s safe to allow further configuration changes as soon as the Cnew entry
has been committed. Unfortunately, this decision does imply that a log entry for a configuration
change can be removed (if leadership changes); in this case, a server must be prepared to fall back
to the previous configuration in its log.

Fun that it specifically mentions how servers do not need to persist commit index in the context of this PR 😄.

That is true for overall linearizable correctness, however I still think this is a useful addition. The raft thesis only (really)considers fully linearizable reads where all reads go through the leader and so there is no downside to a non-leader "accepting" requests before it's fully recovered since it's only handling for requests would be to redirect or forward to the leader. Raft already ensures that a leader doesn't serve requests until all committed logs have been applied. In systems such as Consul and Vault where "stale" reads are explicitly allowed in some cases to increase read scalability, it's certainly not ideal to have a period where a follower has started up and is accepting reads but has not applied all logs it has on it's disk even when it knew they were committed previously.

tl;dr, we're fine here 😄, but we should be careful to preserve the behaviour - even if we only replay up to `commitIndex` on the FSM, we _should_ still process any config changes in the rest of the log after that!

Reply:
@banks Thanks for these lots of comments 😄

Reply:
But,  for the current change, it is possible to apply same configurations twice, although its wouldn't violate the correctness, it still be wasteful. 
So I would reverse the order of `r.recoverFromCommitedLogs()` and "iterating the whole log to see if any logs are config changes", then, if the persisted commit are applied, then we only need to "iterating the ~~whole log~~ `r.logs[max(lastapplied, snapshotIndex) + 1 : ]` to see if any logs are config changes" and that prevent the duplication~